app/helpers/link_to_helper.rb

```diff
@@ -311,4 +308,15 @@ def message_path(message, options = {})
     message_url(message, options.merge(:only_path => true))
   end
 
+  def dom_id(record, prefix = nil)
+    case record
+    when IncomingMessage, OutgoingMessage
+      param_key = record_or_class.class.to_s.gsub('Message', '').downcase
+    else
+      return super
```

Comment:
On one request (a public request made by pro that's part of a batch) I get an error here:


```
ActionView::Template::Error (undefined local variable or method `record_or_class' for #<ActionView::Base:0x000000000414b0>):
    1: <div class="outgoing correspondence js-collapsable" id="<%= dom_id(outgoing_message) %>">
    2:   <%- if cannot?(:read, outgoing_message) %>
    3:     <%= render :partial => 'request/hidden_correspondence', :locals => { :message => outgoing_message } %>
    4:   <%- else %>

app/helpers/link_to_helper.rb:314:in `dom_id'
app/views/request/_outgoing_correspondence.html.erb:1
app/views/request/events/_sent.html.erb:1
app/views/request/_correspondence.html.erb:5
app/views/request/show.html.erb:42
app/views/request/show.html.erb:40
app/controllers/request_controller.rb:100:in `block (3 levels) in show'
app/controllers/request_controller.rb:99:in `block in show'
lib/alaveteli_localization.rb:46:in `with_locale'
app/controllers/request_controller.rb:50:in `show'
app/controllers/application_controller.rb:116:in `record_memory'
lib/deeply_nested_params.rb:15:in `call'
lib/strip_empty_sessions.rb:14:in `call'
```

82d9d2e20d61b1ac370e545fe82350c29c4b9a2b then works, and then 1f3734967592eede73fa98c7965336adff2e464e generates a different error:

```
ActionView::Template::Error (undefined method `to_key' for nil:NilClass
Did you mean?  to_query):
    36: <% end %>
    37:
    38: <%= tag.div incoming_message.get_body_for_html_display(@collapse_quotes),
    39:             id: dom_id(incoming_message.get_main_body_text_part) %>

app/helpers/link_to_helper.rb:320:in `dom_id'
app/views/request/_bubble.html.erb:39
app/views/request/_incoming_correspondence.html.erb:28
app/views/request/events/_response.html.erb:1
app/views/request/_correspondence.html.erb:5
app/views/request/show.html.erb:42
app/views/request/show.html.erb:40
app/controllers/request_controller.rb:100:in `block (3 levels) in show'
app/controllers/request_controller.rb:99:in `block in show'
lib/alaveteli_localization.rb:46:in `with_locale'
app/controllers/request_controller.rb:50:in `show'
app/controllers/application_controller.rb:116:in `record_memory'
lib/deeply_nested_params.rb:15:in `call'
lib/strip_empty_sessions.rb:14:in `call'
```

It is a request that I've been testing destroying the underlying raw email on so there's no content in one of the incoming messages, but it works on develop.

<img width="925" alt="Screenshot 2022-09-27 at 11 04 30" src="https://user-images.githubusercontent.com/282788/192498670-9b375262-b953-4cf5-8353-20001329d8b6.png">

Reply:
Calling `dom_id(nil)` without the override will return the same error so once we come to implement destroying the underlying raw email then we will need to add a conditional here to handle. Should need to handle this case for this PR thought.

Reply:
In the case of 15d1d5864240dc95f855aa5d2793971ed05fdbb9 it looks like there's an error on the `OutgoingMessage`?

```
ActionView::Template::Error (undefined local variable or method `record_or_class' for #<ActionView::Base:0x000000000414b0>):
    1: <div class="outgoing correspondence js-collapsable" id="<%= dom_id(outgoing_message) %>">
```

Having looked a bit more, it looks like the following change is causing the breakage…

```diff
- <%= tag.div incoming_message.get_body_for_html_display(@collapse_quotes) %>
+ <%= tag.div incoming_message.get_body_for_html_display(@collapse_quotes),
+             id: dom_id(incoming_message.get_main_body_text_part) %>
```

…which looks like it's being used to wrap the main body text?

```html
<div class="correspondence_text">
  <div id="attachment-2">
    <p>Dear “Bob”…</p>
  </div>
</div>
```

Why do we need this?

Reply:
Needed to link directly to the attachment, see: https://github.com/mysociety/alaveteli/pull/7327#discussion_r982225290